### PR TITLE
Hive partitioning: Fix preprocessing of CreateDirectories

### PR DESCRIPTION
**Pull request by carlopi**
Fixes https://github.com/duckdblabs/duckdb-internal/issues/588 improving on https://github.com/duckdb/duckdb/pull/9473.

Idea is that we iterate on all global partitions instead of iterating on the local ones.

Thanks @l1t1 for providing the test case.

### CODE REVIEW

**Comment by carlopi** [comment]
I left `HivePartitionedColumnData::GetReverseMap()` call, but that's now unused, if it has no value it can just be removed.

**Comment by carlopi** [comment]
Actually, not yet fixed in a slightly more involved case.

File: src/common/hive_partitioning.cpp [hunk 1]
```diff
@@ -316,6 +316,14 @@ std::map<idx_t, const HivePartitionKey *> HivePartitionedColumnData::GetReverseM
 	return ret;
 }
 
+std::map<idx_t, const HivePartitionKey *> GlobalHivePartitionState::GetReverseMap() {
```
Comment (Mytherin) [inline src/common/hive_partitioning.cpp:319]: Can this be a `const_reference<HivePartitionKey>` instead?

Comment (carlopi) [reply src/common/hive_partitioning.cpp]: This was not needed, so removed, pointer is still there in `HivePartitionedColumnData::GetReverseMap()`, but that I will leave unmodified.

**Comment by Mytherin** [comment]
Thanks!

**Comment by l1t1** [comment]
I tried  the artifacts of https://github.com/duckdb/duckdb/actions/runs/6717806759 , still failed.
```
v0.9.2-dev238 0d83bdfed0
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D create or replace table orders(m int,v int,j int);
D insert into orders select i%12+1,i,j from range(360)t(i),range(1000)s(j);
D copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_m' (FORMAT PARQUET, PARTITION_BY (m));
D copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_y' (FORMAT PARQUET, PARTITION_BY (y));
Error: IO Error: Cannot open file "orders_y\y=2029\data_0.parquet": ????????????·????

D copy (select 2000+(v//12)y,m,v,j from orders) TO 'orders_ym' (FORMAT PARQUET,PARTITION_BY (y,m));
Error: IO Error: Cannot open file "orders_ym\y=2015\m=3\data_1.parquet": ????????????·????

D .system tree orders_y

D:\DUCKDB\ORDERS_Y
├─y=2000
├─y=2001
├─y=2002
├─y=2003
├─y=2004
├─y=2005
├─y=2006
├─y=2007
├─y=2008   -------------------------------no y=2009
├─y=2010
├─y=2011
├─y=2012
├─y=2013
├─y=2014
├─y=2015
├─y=2016
├─y=2017
├─y=2018
├─y=2019
├─y=2020
├─y=2021
├─y=2022
├─y=2023
├─y=2024
├─y=2025
├─y=2026
├─y=2027
└─y=2028
--------------------------------------------no y=2029
D .system tree orders_ym/y=2015

D:\DUCKDB\ORDERS_YM\Y=2015
├─m=1
└─m=2
--------------------------------------------no m=3 and above
```

**Comment by carlopi** [comment]
@l1t1: Thanks for double checking on this, but you mentioned what I think is an outdated link, could you give either https://github.com/duckdb/duckdb/actions/runs/6719943007?pr=9535 (note the link you cited ended with [6717806759](https://github.com/duckdb/duckdb/actions/runs/6717806759)) or nightly builds like https://github.com/duckdb/duckdb/actions/runs/6726710145 ?

Thanks a lot

**Comment by l1t1** [comment]
@carlopi

> @l1t1: Thanks for double checking on this, but you mentioned what I think is an outdated link, could you give either https://github.com/duckdb/duckdb/actions/runs/6719943007?pr=9535 (note the link you cited ended with [6717806759](https://github.com/duckdb/duckdb/actions/runs/6717806759)) or nightly builds like https://github.com/duckdb/duckdb/actions/runs/6726710145 ?
> 
> Thanks a lot

it DOES fixed in https://github.com/duckdb/duckdb/actions/runs/6726710145, thanks

**Comment by carlopi** [comment]
Thanks for checking!
Note that the test https://github.com/duckdb/duckdb/blob/main/test/sql/copy/parquet/parquet_hive2.test is the one you provided, so will be checked moving further.

**Comment by bucweat** [comment]
Ran my test case and results look as expected. Thanks :-)

```
┌─────────────────┬────────────┐
│ library_version │ source_id  │
│     varchar     │  varchar   │
├─────────────────┼────────────┤
│ v0.9.2-dev242   │ e807b416e8 │
└─────────────────┴────────────┘
```

**Pull request merged**